File: ratio-longest-winning-losing-streak/longestrun.py
from decimal import Decimal, getcontext
from scipy.special import comb

FLIPS = [10, 100]
SUCCESS_RATES = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]


def biased_coin_recursion(x, n, p):
    q = 1 - p
    coefs = [[0] * (n + 1) for _ in range(n + 1)]
    for i in range(n + 1):
        for j in range(i + 1):
            if j <= x:
                coefs[i][j] = comb(i, j, exact=True)
            else:
                sumation = 0
                for k in range(x + 1):
                    sumation = sumation + coefs[i - k - 1][j - k]
                coefs[i][j] = sumation
    prob = Decimal(0)
    getcontext().prec = 20
    for i in range(len(coefs)):
        num1 = Decimal(p) ** Decimal(i) * Decimal(q) ** Decimal(n - i)
        prob = prob + (coefs[len(coefs) - 1][i] * num1)
    return prob


for j, flips in enumerate(FLIPS):
    for i, success_rate in enumerate(SUCCESS_RATES):
        fail_rate = 1 - success_rate
        win_streak_cum_probs = []
        loss_streak_cum_probs = []
        for max_streak in range(0, flips + 1):
            win_streak_prob = biased_coin_recursion(max_streak, flips, success_rate)
            loss_streak_prob = biased_coin_recursion(max_streak, flips, fail_rate)
            win_streak_cum_probs.append(win_streak_prob)
            loss_streak_cum_probs.append(loss_streak_prob)

        # win_streak_cum_probs and loss_streak_cum_probs represent the probability at each index i that the longest streak for the respective
        # outcome does not exceed i. From these, we must extract the probability for each i, that the longest streak for the respective outcome
        # is exactly i. This is done by subtracting the cumulative probability at i-1 from the cumulative probability at i.
        win_streak_exact_probs = [win_streak_cum_probs[0]]
        for i in range(1, len(win_streak_cum_probs)):
            win_streak_exact_probs.append(win_streak_cum_probs[i] - win_streak_cum_probs[i - 1])

        loss_streak_exact_probs = [loss_streak_cum_probs[0]]
        for i in range(1, len(loss_streak_cum_probs)):
            loss_streak_exact_probs.append(loss_streak_cum_probs[i] - loss_streak_cum_probs[i - 1])

        expected_ratio = 0
        for k, win_streak_prob in enumerate(win_streak_exact_probs):
            for l, loss_streak_prob in enumerate(loss_streak_exact_probs):
                expected_ratio += (
                    win_streak_prob * loss_streak_prob * Decimal(k / max(l, 1))
                )

        print(f"{flips=}, {success_rate=}, {expected_ratio=}")

File: ratio-longest-winning-losing-streak/longestrun.cpp
#include <gmpxx.h>
#include <iomanip>
#include <iostream>
#include <vector>

using namespace std;

mpf_class biased_coin_recursion(int x, int n, double floatp) {
  double floatq = 1 - floatp;

  mpf_class p(floatp, 500);
  mpf_class q(floatq, 500);
  vector<vector<mpz_class>> coefs(n + 1, vector<mpz_class>(n + 1, 0));

  for (int i = 0; i <= n; ++i) {
    for (int j = 0; j <= i; ++j) {
      if (j <= x) {
        mpz_bin_uiui(coefs[i][j].get_mpz_t(), i, j);
      } else {
        mpz_class summation(0);
        for (int k = 0; k <= x; ++k) {
          if (i - k - 1 >= 0 && j - k >= 0) {
            summation += coefs[i - k - 1][j - k];
          }
        }
        coefs[i][j] = summation;
      }
    }
  }

  mpf_class prob(0, 500);
  for (int i = 0; i <= n; ++i) {
    mpf_class term(0, 500);
    mpf_class first(0, 500);
    mpf_class second(0, 500);
    mpf_pow_ui(first.get_mpf_t(), p.get_mpf_t(), i);
    mpf_pow_ui(second.get_mpf_t(), q.get_mpf_t(), (n - i));
    mpf_set_z(term.get_mpf_t(), coefs[n][i].get_mpz_t());
    term *= first;
    term *= second;
    prob += term;
  }
  return prob;
}

int main() {
  vector<int> FLIPS = {10, 100, 300};
  vector<double> SUCCESS_RATES = {0.05, 0.1, 0.25, 0.3, 0.5, 0.7, 0.9, 0.95};

  for (int flips : FLIPS) {
    for (double success_rate : SUCCESS_RATES) {
      double fail_rate = 1 - success_rate;
      vector<mpf_class> win_streak_cum_probs(flips + 1);
      vector<mpf_class> loss_streak_cum_probs(flips + 1);

      for (int max_streak = 0; max_streak <= flips; ++max_streak) {
        win_streak_cum_probs[max_streak] =
            biased_coin_recursion(max_streak, flips, success_rate);
        loss_streak_cum_probs[max_streak] =
            biased_coin_recursion(max_streak, flips, fail_rate);
      }

      // Calculate exact probabilities
      vector<mpf_class> win_streak_exact_probs(flips + 1);
      vector<mpf_class> loss_streak_exact_probs(flips + 1);

      win_streak_exact_probs[0] = win_streak_cum_probs[0];
      loss_streak_exact_probs[0] = loss_streak_cum_probs[0];

      for (int i = 1; i <= flips; ++i) {
        win_streak_exact_probs[i] =
            win_streak_cum_probs[i] - win_streak_cum_probs[i - 1];
        loss_streak_exact_probs[i] =
            loss_streak_cum_probs[i] - loss_streak_cum_probs[i - 1];
      }

      // Calculate expected ratio
      mpf_class expected_ratio(0, 500);
      for (int k = 0; k <= flips; ++k) {
        for (int l = 0; l <= flips; ++l) {
          if (l != 0) {
            mpf_class term = win_streak_exact_probs[k] *
                             loss_streak_exact_probs[l] *
                             mpf_class(double(k) / l, 500);
            expected_ratio += term;
          }
        }
      }

      cout << "flips=" << flips << ", success_rate=" << success_rate
           << ", expected_ratio=" << expected_ratio << endl;
    }
  }

  return 0;
}
